targets/TARGET_NORDIC/TARGET_NRF5x/flash_api.c

```diff
@@ -38,82 +38,167 @@
-        return -1;
+    (void)(obj);
+
+    /* Check that data pointer is not NULL. */
+    ret_code_t result = NRF_ERROR_NULL;
+
+    if (data) {
+
+        /* nrf_fstorage_write only accepts word aligned input buffers.
+         * Cast data pointer to word pointer and read word into temporary word variable
+         * which should be word aligned. This should be safe on Cortex-m4 which is able
+         * to load unaligned data.
+         */
+        const uint32_t *data_word = (const uint32_t *) data;
+
+        /* Loop through input buffer 4 bytes at a time. */
+        for (size_t index = 0; index < (size / 4); index++) {
+
+            /* Load 4 bytes into temporary variable. */
+            uint32_t word = data_word[index];
+
+            /* Setup stop watch for timeout. */
+            uint32_t start_us = lp_ticker_read();
+            uint32_t now_us = start_us;
+
+            /* Retry if flash is busy until timeout is reached. */
+            do {
+                /* Write one word at a time. */
+                result = nrf_fstorage_write(&nordic_fstorage, address + 4 * index, &word, 4, NULL);
+
+                /* Read timeout timer. */
+                now_us = lp_ticker_read();            
+
+            /* Loop if queue is full or until time runs out. */
+            } while (((now_us - start_us) < PAGE_ERASE_TIMEOUT_US) &&
+                     (result == NRF_ERROR_NO_MEM));
+
+            /* Break loop if write command wasn't accepted. */
+            if (result != NRF_SUCCESS) {
+                break;
+            }
+        }
     }
-#endif
-    /* We will use *_words function to speed up flashing code. Word means 32bit -> 4B
-     * or sizeof(uint32_t).
-     */
-    nrf_nvmc_write_words(address, (const uint32_t *) data, (size / sizeof(uint32_t)));
-    return 0;
+
+    /* Convert Nordic SDK error code to mbed HAL. */
+    return (result == NRF_SUCCESS) ? 0 : -1;
 }
 
 uint32_t flash_get_size(const flash_t *obj)
 {
     (void)(obj);
-    /* Just count flash size. */
-    return NRF_FICR->CODESIZE * NRF_FICR->CODEPAGESIZE;
+
+    return nordic_fstorage.end_addr - nordic_fstorage.start_addr;
+    ;
 }
 
 uint32_t flash_get_sector_size(const flash_t *obj, uint32_t address)
 {
     (void)(obj);
-    /* Test if passed address is in flash space. */
-    if (address < flash_get_size(obj)) {
-        return NRF_FICR->CODEPAGESIZE;
+
+    /* Check if passed address is in flash area. Note end_addr is outside valid range. */
+    if ((address >= nordic_fstorage.start_addr) && (address < nordic_fstorage.end_addr)) {
+        return nordic_fstorage.p_flash_info->erase_unit;
     }
-    /* Something goes wrong, return invalid size error code. */
+
+    /* Return invalid size if request is outisde flash area. */
     return MBED_FLASH_INVALID_SIZE;
 }
 
 uint32_t flash_get_page_size(const flash_t *obj)
 {
     (void)(obj);
-    return NRF_FICR->CODEPAGESIZE;
+
+    /* Return minimum writeable page size. */
+    return 4;
```

Comment:
better to use a pre defined value instead of raw 4

Reply:
Changed to #define WORD_SIZE_IN_BYTES.